oeps/conventional-commits.rst

```diff
@@ -0,0 +1,77 @@
+####################
+Conventional Commits
+####################
+
+Abstract
+********
+
+Commits should be clearly labeled for their effect on consumers of the repository.  To do this, we adopt the `Conventional Commits`_ guidelines, which detail structured commit messages that clarify the impact of each commit.
+
+This is part of our Change Transparency initiative.
+
+Spec
+****
+
+We are adopting the `Conventional Commits`_ spec, with our own adjustments.
+
+.. note:: Many people are familiar with the `Angular commit message format`_, which uses conventional commits and inspired the Conventional Commits spec.  We are not adopting the Angular rules.
+
+Commit messages have these parts::
+
+    <type>: <subject>
+
+    <body>
+
+    <footer>
+
+Type
+====
+
+We use these commit types:
+
+* **build**: anything to do with building or releasing the repo: Makefile, tox.ini, Travis, Jenkins, GitHub actions, and so on.
+* **chore**: a repetitive task such as updating requirements or translations.
+* **docs**: documentation-only changes. Documentation changes for other types of work can be included in those commits.
+* **feat**: a new feature, or a change to an existing feature. Anything that changes the set of features.
```

Comment:
One thing that's not clear in the CC spec is how to label generic code "improvements" that aren't exactly _bug_ fixes, and don't introduce new or deprecate existing features, but just... improve them. Stability, security, usability, etc. Performance improvement is already called out as a separate type, but that still leaves this intermediate space ambiguous. (I've already run into this when upgrading a cryptographic signer to a newer, more trusted algorithm.) Sometimes these are user visible, sometimes not.

Would we err on the side of `feat` for these improvements? Or consider any improvement to be a `fix`? Or should there be a different type?

Reply:
@timmc-edx, from the point of view of someone looking to use commit messages as raw material for release notes, I would very much prefer seeing such improvements under `feat`.

Reply:
I think this is an interesting point.  User-visible changes to features feel different than improved stability or better crypto.  I'm not sure where I would put them.  I think I would tend toward "fix", since they are making an improvement. The reason you made the change was because the existing features will work better than they did before.

Reply:
Maybe the way to explain this is:  "feat" is for changes that would get mentioned in user docs.  It's about what the software does.  "fix" is for changes to improve how the software does it.  

Or did that only make it worse?

Reply:
Anecdote: I've made a bajillion commits recently that amount to changing `student.models` to `common.djangoapps.student.models`, and have been wondering which commit type I shoud be using. The only visible change is that a deprecation warning is removed. It's not necessarily a bugfix, but it feels very generous to call that a `feat`.

Reply:
@kdmccormick Would that be a `refactor` type?

Reply:
Another anecdote that comes to mind are [the XModule-to-XBlock conversions](https://github.com/edx/edx-platform/pull/25965). The goal of the project is to remove our dependency on `XModule` while having no user-facing impact.

Not exactly a `fix`, but not a `feat` either.

Reply:
[suggestion] There could be a commit type called `quality`, which would encapuslate this grey area of code-improvements-that-aren't-quite-bugfixes-or-features category. The `style` type could be consolidated into it as well.

Reply:
> @kdmccormick Would that be a refactor type?

@nasthagiri , that's a good point. Maybe these grey areas are all just instances of refactoring.

Reply:
@timmc-edx It would be useful to have specific examples to consider.

- performance improvement -> `perf`
- security fix -> 'fix`
- usability -> `feat`
- other types of code changes -> `refactor`

Reply:
I would call a usability improvement a "fix", not a "feat".  It doesn't change what the software does, it changes how it does it.

Reply:
> I would call a usability improvement a "fix", not a "feat". It doesn't change what the software does, it changes how it does it.

There are multiple benefits of having all user-facing changes designated as "feat" or the like:

- It signals to PR review managers that a Product review is required. _(All other commit-types do not require product review.)_
- It signals the possibility of an a11y and UX review.
- Per @arbrandes's point above, it signals inclusiveness in release notes.

Reply:
I'm having a hard time describing every user-facing change as a feature.  If we want fixes to be included in the release notes, then we should read the "fix" commits to include them.  How do other people feel about it?

Reply:
Here are a few tweaks that I think could satisfy some of the concerns in this thread:

* Grouping according to who the commit should interest (no one, developers, everyone). `feat` is also split into `api` (maybe there is a better name?) and `ux` for that.
* `-major` suffix means the developer making the change does the work of figuring out the significance and not whoever is compiling release notes but there are not too many different types to worry about.

Thoughts?

------

Types that should never introduce a breaking change or a UX change:
* **chore**: a repetitive task such as updating requirements or translations.
* **docs**: documentation-only changes. Documentation changes for other types of work can be included in those commits.
* **style**: improve the styling of the code.
* **test**: test-only changes. Adds missing tests or corrects existing tests. Tests accompanying other types of changes go into those commits.

Types for changes that impact developers but not the UX (any of them could or could not be breaking):
* **api-major**: new APIs, significant changes to APIs or other changes that impact developers that are worth listing in release notes.
* **api**: code reorganization, smaller API changes or fixes.
* **build**: anything to do with building or releasing the repo: Makefile, tox.ini, Travis, Jenkins, GitHub actions, and so on.
* **performance**: performance improvement. Significant user visible changes should be marked `ux` instead.

Type for changes that impact the UX:
* **ux-major**: a new feature or a change to UX that is significant and worth listing in the release notes.
* **ux**: a fix or a minor change to UX.

In case of reverts, if all the commits in the revert are of one type use that type. Otherwise:
* **revert**: undo a set of previous changes of different type. The previous commit message, including the type prefix, is included.

Reply:
Works for me. ship it!

Reply:
This is a departure from the existing conventional commit spec and common usage.  I think we want to stick with the spec as much as we can.  No one else likes the idea of using scopes to address the shortcomings of the types?

Reply:
What would using `scopes` look like? Would it over-complicate things, when @symbolist's proposal seems relatively simple and sufficient?

Reply:
As a future evolution, I'm thinking we would have folks include what area/app of the platform it affects (since that's useful to know at a glance). For example "grades", "courseware", "certificates", etc.  So that would also add another layer of specificity.

Reply:
Under the new proposal my example of a trivial refactoring (`student.models` -> `common.djangoapps.student.models`) becomes a change of type `api:`, which doesn't feel right.

I am leaning towards the original proposal, in which:
* `refactor` doesn't change behavior
* `fix` fixing an existing behavior  (ux, HTTP API, or otherwise)
* `feat` adds new behavior (ux, HTTP API, or otherwise)

Since the frontend replatforming, our direction is that edx-platform is getting fewer and fewer `ux` type changes, with the hope that the edx-platform UIs will eventually go away. Given that, I believe that the `feat`s of edx-platform (or any of our microservices) are naturally going to be HTTP API changes; splitting out `ux` and `api` doesn't seem necessary.

Reply:
Catching up on the whole thread here -

It seems problematic to me to deviate significantly from the established types that many projects already use, and which are supported out of the box by tools like semantic-release.  And for what it's worth, it would also be different than what we're already doing in all our frontend repositories.

There's a lot of flexibility in the original proposal, which is based on prior art from other projects/tools.

1. Adding color to the types is what `scope` is for, and any given commit WILL have its commit message associated with it still to add even more context.  

2. If something doesn't fit exactly into fix vs. feat vs. non-version-impacting-types, a prefix can be chosen by the effect an engineer thinks it should have on the software version.  When viewed in the release notes, a mildly mis-typed commit with a good message can presumably be forgiven and/or understood anyway with little impact.

Point being, between the existing types, adding a scope, and the message itself, plus any other identifiers we want to decorate our messages with, it seems like we have _plenty_ of nuance to work with without inventing our own standard types.

Reply:
The tweaks I suggested were based on the the assumption that it is useful to be able to tell programmatically:

* If a change only impacts developers only or end users as well.
* If it a large enough change to be worth highlighting in the release notes. Given the high volume of changes in edX a list of all the fix and feature commits is still going to be very long for anyone to process.

So the first question is whether it is useful to be able to programmatically differentiate in one or both of the above? If it is not then yeah we don't need to add anything beyond the base spec and only need to decide on the conventions for the ambiguous cases that were discussed earlier in the thread.

If however it would be useful to programmatically differentiate along either of the two axis, are there any other proposals? Though it does not look like we can use scopes since the spec says:

> A scope MAY be provided after a type. A scope MUST consist of a noun describing a section of the codebase surrounded by parenthesis, e.g., fix(parser):

Reply:
I would agree with @davidjoy  and @nedbat about sticking closely to the spec.  I think especially at the beginning since we haven't seen it in action yet.  I think what @nasthagiri summarized is good enough for now and we can iterate if we find a lot of friction with this initial approach.


>     * performance improvement -> `perf`
>     * security fix -> `fix`
>     * usability -> `feat`
>     * other types of code changes -> `refactor`